internal/pkg/service/buffer/storage/local/reader/reader.go

```diff
@@ -0,0 +1,174 @@
+	ref, exists := v.addReaderFor(slice.SliceKey)
+	if exists {
+		return nil, errors.Errorf(`reader for slice "%s" already exists`, slice.SliceKey.String())
+	}
+
+	// Close resources on a creation error
+	var file File
+	var chain *readchain.Chain
+	defer func() {
+		if err == nil {
+			// Update reference
+			ref.SliceReader = r
+		} else {
+			// Close resources
+			if chain != nil {
+				_ = chain.Close()
+			} else if file != nil {
+				_ = file.Close()
+			}
+			// Unregister the reader
+			v.removeReader(slice.SliceKey)
+		}
+	}()
+
+	// Open file
+	dirPath := filesystem.Join(v.path, slice.LocalStorage.Dir)
+	filePath := filesystem.Join(dirPath, slice.LocalStorage.Filename)
+	file, err = v.config.fileOpener(filePath)
+	if err == nil {
+		logger.Debug("opened file")
+	} else {
+		logger.Error(`cannot open file "%s": %s`, filePath, err)
+		return nil, err
+	}
+
+	// Init reader and chain
+	chain = readchain.New(logger, file)
+	reader := &sliceReader{readChain: chain, sliceKey: slice.SliceKey, dirPath: dirPath, filePath: filePath}
+
+	// Unregister the reader on close
+	chain.AppendCloseFn("unregister", func() error {
+		v.removeReader(slice.SliceKey)
+		return nil
+	})
+
+	// Compare local and staging compression
+	if slice.LocalStorage.Compression.Type == slice.StagingStorage.Compression.Type {
+		// Local and staging compression types are same.
+		// Return the chain with only the *os.File inside.
+		//
+		// See the Chain.UnwrapFile method, to get the underlying *os.File.
+		//
+		// This is preferred way, it enables internal Go optimization and "zero CPU copy" to be used,
+		// read more about "sendfile" syscall for details.
+		return reader, nil
+	}
+
+	// Decompress the file stream on-the-fly, when reading.
+	if slice.LocalStorage.Compression.Type != compression.TypeNone {
+		_, err := chain.PrependReaderOrErr(func(r io.Reader) (io.Reader, error) {
+			return compressionReader.New(r, slice.LocalStorage.Compression)
+		})
+		if err != nil {
+			return nil, errors.Errorf(`cannot create compression reader: %w`, err)
+		}
+	}
+
+	// Compress the file stream on-the-fly, when reading.
+	if slice.StagingStorage.Compression.Type != compression.TypeNone {
+		// Convert compression writer to a reader using pipe
+		pipeR, pipeW := io.Pipe()
+		compressionW, err := compressionWriter.New(pipeW, slice.StagingStorage.Compression)
+		if err != nil {
+			return nil, errors.Errorf(`cannot create compression writer: %w`, err)
+		}
+		chain.PrependReader(func(r io.Reader) io.Reader {
+			// Copy: raw bytes (r) -> compressionW -> pipeW -> pipeR
+			go func() {
+				var err error
+				if _, copyErr := io.Copy(compressionW, r); copyErr != nil {
+					err = copyErr
+				}
+				if closeErr := compressionW.Close(); err == nil && closeErr != nil {
+					err = closeErr
+				}
+				_ = pipeW.CloseWithError(err)
+			}()
+			return pipeR
+		})
+	}
+
+	return reader, nil
+}
```

Comment:
V pripade potrebny mozeme kompresiu zmenit:
- napr. `None` v local storage -> `GZIP` v staging storage, to by davalo zmysel, ak by sme mali rychli disk, a chceli data komprimovat az pri uploade, ... napr. preto, lebo po vacsich blocks je lepsie `compression ratio` (ak by jednotlive records boli velmi male).